### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by SamoylenkVi**

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -30,5 +30,6 @@ module.exports = {
     'no-underscore-dangle': 'off',
     'class-methods-use-this': 'off',
     'import/prefer-default-export': 'off',
+    'react/react-in-jsx-scope': 'off',
```
Comment (Pavel-Liteiniy) [inline .eslintrc.js]: [react/react-in-jsx-scope'](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/react-in-jsx-scope.md#when-not-to-use-it)

> If you are not using JSX, or if you are setting React as a global variable.

> If you are using the [new JSX transform from React 17](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#removing-unused-react-imports), you should disable this rule by extending [react/jsx-runtime](https://github.com/jsx-eslint/eslint-plugin-react/blob/8cf47a8ac2242ee00ea36eac4b6ae51956ba4411/index.js#L165-L179) in your eslint config (add "plugin:react/jsx-runtime" to "extends").

Comment (Pavel-Liteiniy) [reply .eslintrc.js]: So than there is no need to make `import React from 'react'` in all other files - delete everywhere

File: src/components/film-card.jsx [hunk 2]
```diff
@@ -1,28 +1,61 @@
...
+import cn from 'classnames';
 import { REGEX_ALT } from '../constants';
 import { createAltText } from '../utils/film-card';
+import { VideoPlayer } from './video-player.jsx';
 
 export const FilmCard = ({
-  name, previewImage, id, onMouseEnter,
+  name, previewImage, previewVideoLink, id, onMouseEnter,
 }) => {
+  const videoRef = React.createRef();
```
Comment (Pavel-Liteiniy) [inline src/components/film-card.jsx]: why not useRef?

File: src/components/film-card.jsx [hunk 3]
```diff
@@ -1,28 +1,61 @@
...
+    if (onMouseEnter) {
+      onMouseEnter(id);
+    }
+    setVideo(!isVideo);
+  };
+
+  useEffect(() => {
+    if (isVideo && videoRef.current) {
+      videoRef.current.play().catch(() => {
+        setVideoError(!isVideoError);
```
Comment (Pavel-Liteiniy) [inline src/components/film-card.jsx]:
```suggestion
        setVideoError(true);
```

File: src/components/film-card.jsx [hunk 4]
```diff
@@ -1,28 +1,61 @@
...
+      });
+    }
+  }, [isVideo]);
 
   return (
-    <article onMouseEnter={() => onMouseEnter(id)} className="small-movie-card catalog__movies-card">
+    <article onMouseEnter={playVideo} onMouseLeave={playVideo} className="small-movie-card catalog__movies-card">
```
Comment (Pavel-Liteiniy) [inline src/components/film-card.jsx]: why do you put same cb in opposite handlers?

File: src/components/film-card.jsx [hunk 5]
```diff
@@ -1,28 +1,61 @@
...
       <div className="small-movie-card__image">
-        <img src={previewImage} alt={altText} width="280" height="175" />
+        {isVideo
+          ? <VideoPlayer src={previewVideoLink} ref={videoRef} />
+          : <img src={previewImage} alt={altText} width="280" height="175" />
+        }
+        {isVideoError && <img src={previewImage} alt={altText} width="280" height="175" />}
```
Comment (Pavel-Liteiniy) [inline src/components/film-card.jsx]: May it happen that there is isVideo and isVideoError in the same time?

File: src/components/film-details-information/film-details-information.jsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          <p className="movie-card__details-item">
+            <strong className="movie-card__details-name">Director</strong>
+              <span className="movie-card__details-value">{director}</span>
+          </p>
+          <p className="movie-card__details-item">
+            <strong className="movie-card__details-name">Starring</strong>
+            <span className="movie-card__details-value">
+              {starring.map((item, index) => {
+                if (index === starring.length - 1) {
+                  return <>
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-information/film-details-information.jsx:20]: use Fragment and add `key` attribute

File: src/components/film-details-information/film-details-information.jsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                    {item}
+                  </>;
+                }
+                return <>
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-information/film-details-information.jsx:24]: so difficult to read.

Please use this 


```suggestion
                return (
                  <Fragment key={id}>
                    {item}, <br></br>
                  </Fragment>
                );
```

File: src/components/film-details-information/helper.js
Comment (Pavel-Liteiniy) [inline src/components/film-details-information/helper.js:1]: helper**s**.js

File: src/components/film-details-navigation/details-navigation.jsx [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+export const DetailsNavigation = ({ film }) => {
+  const [currentButton, setCurrentButton] = useState(NAVIGATION_BUTTONS.DETAILS);
+
+  const handlerActiveButton = (evt) => {
+    const buttonName = evt.target.getAttribute('data-name');
+    setCurrentButton(buttonName);
+  };
+
+  useEffect(() => {
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-navigation/details-navigation.jsx]: why?

File: src/components/film-details-navigation/details-navigation.jsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+          { Object.values(NAVIGATION_BUTTONS).map((button) => (
+            <>
+              <li className="movie-nav__item">
+                <button
+                  key={button}
+                  data-name={button}
+                  onClick={handlerActiveButton}
+                  className={cn('movie-nav__link', { 'movie-nav__link--active': currentButton === button })}
+                >
+                  {ucFirst(button)}
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-navigation/details-navigation.jsx:29]: what does `ucFirst` mean?

File: src/components/film-details-navigation/details-navigation.jsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+  useEffect(() => {
+  }, [currentButton]);
+
+  return (
+    <>
+      <nav className="movie-nav movie-card__nav">
+        <ul className="movie-nav__list">
+          { Object.values(NAVIGATION_BUTTONS).map((button) => (
+            <>
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-navigation/details-navigation.jsx:21]: You should add key to the root component https://react.dev/learn/rendering-lists

here it's `Fragment`

File: src/components/film-details-navigation/details-navigation.jsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+                  className={cn('movie-nav__link', { 'movie-nav__link--active': currentButton === button })}
+                >
+                  {ucFirst(button)}
+                </button>
+              </li>
+            </>
+          ))}
+        </ul>
+      </nav>
+      <FilmDetailsContent film={film} state={currentButton} />
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-navigation/details-navigation.jsx:36]: let's rename prop `state` to `currentButton` for example

File: src/components/film-details-navigation/helper.js [hunk 12]
```diff
@@ -0,0 +1 @@
+export const ucFirst = (str) => str[0].toUpperCase() + str.slice(1);
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-navigation/helper.js:1]: strange name for function - what does it mean?

File: src/components/film-details-navigation/details-navigation.jsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    </>
+  );
+};
+
+DetailsNavigation.propTypes = {
+  film: PropTypes.objectOf(),
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-navigation/details-navigation.jsx:42]: object of what?

File: src/components/film-details-overview/helper.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
+const RATING__LEVEL = {
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-overview/helper.js]: let's move to `constants`

File: src/components/film-details-overview/helper.js [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+  BAD: 'Bad',
+  NORMAL: 'Normal',
+  GOOD: 'Good',
+  VERY_GOOD: 'Very good',
+  AWESOME: 'Awesome',
+};
+
+const STARRING_MAX = 3;
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-overview/helper.js]: let's move to `constants`

File: src/components/film-details-overview/helper.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+export const calculateRatingLevel = (rating) => {
+  if (rating < 3) {
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-overview/helper.js]: 3, 5, 8 etc. - magic numbers

File: src/components/film-details-page/helper.js [hunk 17]
```diff
@@ -0,0 +1,5 @@
+const SIMILAR_FILMS_MAX = 4;
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-page/helper.js]: let's move to `constants`

File: src/components/film-details-review/filmDetailsReview.jsx [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+export const FilmDetailsReview = ({ comments }) => {
+  const commentComponents = comments.map((comment) => (
+    <FilmReview
+      comment={comment} key={comment.id}
+    />
+  ));
+
+  return (
+    <>
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-review/filmDetailsReview.jsx]: why do you use fragment here?

File: src/components/film-details-review/filmDetailsReview.jsx [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+        <div className="movie-card__reviews-col">
+          {commentComponents.slice(3, 6)}
+        </div>
+      </div>
+    </>
+  );
+};
+
+FilmDetailsReview.propTypes = {
+  comments: PropTypes.objectOf(),
```
Comment (Pavel-Liteiniy) [inline src/components/film-details-review/filmDetailsReview.jsx]: object of what?

**Review by Pavel-Liteiniy — CHANGES_REQUESTED** [review]

File: src/components/film-card.jsx (hunk 2 above)
Comment (SamoylenkVi) [reply src/components/film-card.jsx]: fixed

File: src/components/film-details-overview/helper.js (hunk 15 above)
Comment (SamoylenkVi) [reply src/components/film-details-overview/helper.js]: fixed

File: src/components/film-details-overview/helper.js (hunk 16 above)
Comment (SamoylenkVi) [reply src/components/film-details-overview/helper.js]: fixed

File: src/components/film-details-page/helper.js (hunk 17 above)
Comment (SamoylenkVi) [reply src/components/film-details-page/helper.js]: fixed

File: src/components/film-details-review/filmDetailsReview.jsx (hunk 18 above)
Comment (SamoylenkVi) [reply src/components/film-details-review/filmDetailsReview.jsx]: fixed

File: .eslintrc.js (hunk 1 above)
Comment (SamoylenkVi) [reply .eslintrc.js]: fixed

File: src/components/film-card.jsx (hunk 3 above)
Comment (SamoylenkVi) [reply src/components/film-card.jsx]: fixed

File: src/components/film-card.jsx (hunk 4 above)
Comment (SamoylenkVi) [reply src/components/film-card.jsx]: fixed

File: src/components/film-card.jsx (hunk 5 above)
Comment (SamoylenkVi) [reply src/components/film-card.jsx]: fixed

File: src/components/film-details-review/filmDetailsReview.jsx (hunk 19 above)
Comment (SamoylenkVi) [reply src/components/film-details-review/filmDetailsReview.jsx]: fixed

**Comment by Pavel-Liteiniy** [comment]
Approved, good job! 👍

**Pull request merged**